Postgres database setup with pgadmin running on Docker

jcargoncalves opened the pull request:
Added support for a postgresql database together with pgadmin for development and testing. Each running on a separate docker container.

Docker images used were:
postgres (setupdb)
dpage/pgadmin4 (pgadmin)

- To run the db:

npm run setupdb (Default port 5432)

- After the db is set, a pgadmin instance can be started:

npm run pgadmin (Default port 15432)

All ports are mapped to the host running the script. Running the setupdb script also creates a ./postgres-data directory that is mounted on the setupdb docker container, in order to enable data persistence for tables, etc.

Pgadmin config files are already set at the ./pgadmin subdirectory (server.json and pgpass). Additional configuration such as environment variables are present at ./.env.db

jcargoncalves committed:
Docker compose setup

- Added two new docker-compose files for both the pgadmin and the postgres db containers.
- Added a new directory "pgadmin" with the pgadmin config files

jcargoncalves committed:
Updated .gitignore file for setupdb

Add a new ignore line for the temporary directory where the db data (tables,etc...) will persist.

jcargoncalves committed:
Added the npm lines for db and pgadmin

- Added two new npm script lines for both postgres db and pgadmin running on separate docker instances.

jcargoncalves committed:
Added DB environment variables

jcargoncalves committed:
Update docker and env files

Updated files to properly use the environment variables as configuration parameters.

claudenirmf commented on .gitignore:
```diff
@@ -10,6 +10,9 @@
 /build
 
+# db volume
+.postgres-data
```
Why can you ignore the postgres volume, but not the pgadmin one?

jcargoncalves replied:
The pgadmin directory had the servers.json needed for pgadmin to run with the correct db configuration "out of the box". I have simplified in my new version of the code, removing the directory and generating the json files dynamically from the environment variables. I will comment in details on my new commits

claudenirmf commented on pgadmin/pgpass:
```diff
@@ -0,0 +1 @@
+pgdb:5432:test_db:root:root
```
What does this file do?

claudenirmf replied:
Can you have this on the `.env` and provide the value to `pgAdmin` through the `docker-compose._pgadmin.yml` file?

claudenirmf replied:
By the way, if we cannot have this on an environment file, it will quite annoying changing the postgres port (default `5432`) because we'll have to change multiple files (going through that right now myself).

jcargoncalves replied:
I've removed the need of this file. I will address this issue on my new commits.

claudenirmf commented on pgadmin/servers.json:
```diff
@@ -0,0 +1,14 @@
+{
```
What does this file do?

claudenirmf replied:
Can you have this on the `.env` and provide the value to `pgAdmin` through the `docker-compose._pgadmin.yml` file?

jcargoncalves replied:
The servers.json file is necessary for the docker image we are using for pgadmin (dpage/pgadmin4). It is used to load the postgresql db configuration (hostname, port,etc) automatically when the docker container is loaded.

claudenirmf commented on package.json:
```diff
@@ -7,7 +7,9 @@
     "server": "node src/server.js",
-    "test": "jest"
+    "test": "jest",
+    "setupdb": "docker-compose -f docker-compose_setupdb.yml up -d",
```
Do not run it detached. I want it in the foreground to easily interact with the process through the terminal and to terminate it with a `ctrl+C`.

claudenirmf replied:
Also, change the `setupdb` name to just `db`. "Setup" is a bad name here because that is not what you are doing. You are running the db, not setting it up.

claudenirmf replied:
Also, change the name of the file name from `docker-compose_setupdb.yml` to `docker-compose_postgres.yml`, referring to the name of the application, not one of its steps.

claudenirmf commented on package.json:
```diff
@@ -7,7 +7,9 @@
-    "test": "jest"
+    "test": "jest",
+    "setupdb": "docker-compose -f docker-compose_setupdb.yml up -d",
+    "pgadmin": "docker-compose -f docker-compose_pgadmin.yml up -d"
```
Do not run it detached. I want it in the foreground to easily interact with the process through the terminal and to terminate it with a `ctrl+C`.

claudenirmf commented on .env.db:
```diff
@@ -0,0 +1,19 @@
+DB_EXT_PORT=5432
```
Why did you create a new `.env` file? Is it really necessary? Or can you use `.env.example`?

jcargoncalves replied:
In fact, it was just me being extra careful with the repository files. I didn't know if the main application would always run with the db included, as there are two env files (.env.example and .env.test). If this is the default scenario, I can easily copy my environment variables to .env.example.

claudenirmf commented on docker-compose_pgadmin.yml:
```diff
@@ -0,0 +1,13 @@
+    container_name: pgadmin
+    image: dpage/pgadmin4
+    environment:
+      PGADMIN_DEFAULT_EMAIL: "${PGADMIN_LOGIN}"
```
How does `docker-compose` knows where to find these environment variables? Do you need a default `.env` file? Did it get the values from `.env.db`?

jcargoncalves replied:
It gets the variables from .env. The other .env files are just templates to be renamed and copied as .env, in order to run the application.

claudenirmf reviewed (changes requested):
Just went through the file changes. I'll be running the project on my machine now.

claudenirmf commented on pgadmin/servers.json:
```diff
@@ -0,0 +1,14 @@
+        "test": {
+            "Name": "test_db",
+            "Group": "Servers",
+            "Port": 5432,
```
By the way, if we cannot have this on an environment file, it will quite annoying changing the postgres port (default `5432`) because we'll have to change multiple files (going through that right now myself). The same applies to other variables,

claudenirmf committed:
Updates ignored postgres folder and names

jcargoncalves committed:
Add a nodejs script to create a servers.json config file

Added a nodejs script to create a servers.json config file, in order to mke pgadmin able to start with the correct postgresql db setup configuration

jcargoncalves committed:
Changed package.json in order to call pgadmin config script.

Changed package.json in order to call pgadmin config script. It uses an "&&" operand to make it able to generate the config file and run docker-compose afterwards.

jcargoncalves committed:
Update .gitignore with the new files

Update .gitignore with the new files. Added the new pgadmin-servers.json file, as it is generated by the script.

jcargoncalves committed:
Remove pgadmin directory and pgpass

Remove pgadmin directory and pgpass. They are not necessary for the new docker setup.

jcargoncalves committed:
Update docker-compose files and env variables

Update docker-compose files and env variables. Now the postgresql db port and hostname are set based on the variables at .env. Included .env.db as an example to be used for the db operations.

joaogoncalves777 commented:
Two small issues I have found on Linux (Specifically Ubuntu 20.04):

- The pgadmin web interface takes a while to start (Same as reported by claudenirmf on Mac).
- The postgres-data directory's owner is changed to "systemd-coredump" after the docker postgres container mounts the directory on its filesystem.

The first problem seems to be either an issue with the docker image or with pgadmin4 itself. The second is a bug with ubuntu and docker, but none of them interferes with the usage of the db or the web interface.

claudenirmf committed:
Updates environment variables
